### ISSUE TITLE
Nomad should try and schedule enough healthy allocations to meet the count requirement without being bound by broken unhealthy allocations that cant be removed.

### ISSUE THREAD

**Issue by hvindin** [hashicorp/nomad#4862]
### Nomad version
Nomad v0.8.6+ent (306d85c455fe13d67ac788c890aad225863dd531+CHANGES)
(also version 0.8.5)

### Operating system and Environment details
RHEL 7.3 & RHEL 7.4
Docker EE 18.03.3 & Docker EE 18.03.4

### Issue
When a container allocation becomes unhealthy and nomad tries to reschedule it we have observed instances where docker fails to remove the container (which is definitely a docker problem, their RHEL support is a little lacking) and nomad, having failed to stop the running contained just hangs and doesn't attempt to schedule a new one on another host.

I know that I might technically only be declaring in my job file that I want "count = 2" but it would seem logical that this should imply that I actually want 2 *healthy* allocations. In which case it shouldnt matter if nomad and docker tie themselves in knots trying to clean up the unhealthy allocation, I would want nomad to get working on creating a new healthy allocation, possibly on another node if it detects a failure removing the unhealthy allocation.

### Reproduction steps
This is tricky and I'll update here when I figure out how to successfully reproduce, but for now it may be worth considering this a question of nomads intended behavior vs what it does when an allocation cant be removed for some reason.

**Comment by preetapan** [comment]
@hvindin Can you share the output of `nomad alloc status <>` of the container that's in this state. In general if its `ClientStatus` is failed, it will get rescheduled on to another node.

**Comment by stevenscg** [comment]
@hvindin I think I have similar problems on a similar setup as you: Nomad 0.8.6, CentOS 7.5, Docker CE 18.06.1-ce (overlay2 on xfs).

This condition often results in service outages for us and multiple total node failures throughout the cluster.

Do you see something like this when looking at an alloc status??

```
Recent Events:
Time                  Type              Description
2018-11-19T16:19:57Z  Not Restarting    Error was unrecoverable
2018-11-19T16:19:57Z  Driver Failure    failed to start task "api" for alloc "2a77ecf1-bba3-2e77-854e-522a5fa70b72": Failed to create container: Failed to purge container 20cec463f38e8f8de766ab7d53550afb187dddaf284c59d6abfd9dcca97f47dc: API error (409): {"message":"removal of container 20cec463f38e8f8de766ab7d53550afb187dddaf284c59d6abfd9dcca97f47dc is already in progress"}
2018-11-19T16:19:57Z  Restarting        Task restarting in 0s
2018-11-19T16:14:57Z  Killed            Task successfully killed
2018-11-19T16:14:42Z  Killing           Sent interrupt. Waiting 30s before force killing
2018-11-19T16:14:42Z  Restart Signaled  Template: template with change_mode restart re-rendered
2018-11-19T16:14:20Z  Signaling         Task being sent signal user defined signal 2: Template: template re-rendered
2018-11-19T12:31:39Z  Signaling         Task being sent signal user defined signal 2: Template: template re-rendered
2018-11-19T12:14:54Z  Signaling         Task being sent signal user defined signal 2: Template: template re-rendered
2018-11-18T14:10:09Z  Started           Task started by client
```

**Comment by hvindin** [comment]
In looking into this further, the core issue is *definitely* a docker problem in that a job can end up in an unresponsive state because the docker daemon locks up.

Ignoring all the little things that went wrong at some point to end up leaving jobs in this state we have worked out that the _docker healthcheck_ is essentially killing off our containers when a machine gets too busy (it doesn't help that the CPU on the underlying host, we've since discovered, was overallocated up to 600%+, so lots of weirdness starts happening at that point...).

So, ideally it would be _nice_ if nomad were able to figure out when this sort of thing happened and make sure that there were always `$count` healthy allocations available, even if other unhealthy allocation couldn't be removed.

It would also be _nice_ if docker were just that little bit more stable (seriously, search for "healthcheck" in the moby issues... apparently this has been happening for ages and every release since 1.11.1 has "fixed" it....)

But in the short term, it seems that if we kill off the docker healthchecks, and also talk angrily at our "cloud" provider about the absurdity of seeing stats about CPU commit on ESX hosts in the high triple digits then this happens so infrequently that we haven't been able to leave it hung long enough to investigate it because we can't really justify it.

That being said, one thing that would be nice is for nomad to let us turn off native healthchecks at run time so we can run the same container versions but redeploy them with the healthchecks set to `NONE`. So I've created #5310 about that.
